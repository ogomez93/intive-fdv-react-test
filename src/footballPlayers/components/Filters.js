import React from 'react';
import { connect } from 'react-redux';
import { setNameFilterPreview, setPositionFilterPreview, setAgeFilterPreview, applyFilters } from '../actions';
import { POSITIONS } from '../constants';

const positions = POSITIONS.map(position =>
  <option
    value={position}
    key={position}>
    {position}
  </option>
);

const Filters = ({ nameFilterPreview, ageFilterPreview, positionFilterPreview, onNameChange, onPositionChange, onAgeChange, onSearchSubmit }) => {
  let nameInput, positionSelect, ageInput;
  return (
    <form className="my-4 text-xs sm:flex sm:justify-between" onSubmit={ onSearchSubmit }>
      <input
        className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
        onChange={ () => onNameChange(nameInput.value) }
        ref={ node => nameInput = node }
        placeholder="Player Name" />
      <select
        className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4 bg-white"
        onChange={ () => onPositionChange(positionSelect.value) }
        ref={ node => positionSelect = node }
        placeholder="Position">
        <option value="">Position</option>
        { positions }
      </select>
      <input
        className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
        type="number"
        onChange={ () => onAgeChange(ageInput.value) }
        ref={ node => ageInput = node }
        placeholder="Age"
        min="18"
        max="40" />
      <button
        className="border-2 border-black px-6 py-3 w-full sm:w-auto"
        type="submit">
        Search
      </button>
    </form>
  )
};

const mapStateToProps = (state) => {
  return {
    nameFilterPreview: state.footballPlayers.nameFilterPreview,
    ageFilterPreview: state.footballPlayers.ageFilterPreview,
    positionFilterPreview: state.footballPlayers.positionFilterPreview
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onNameChange: (value) => dispatch(setNameFilterPreview(value)),
    onPositionChange: (value) => dispatch(setPositionFilterPreview(value)),
    onAgeChange: (value) => dispatch(setAgeFilterPreview(value)),
    onSearchSubmit: (event) => {
      event.preventDefault();
      dispatch(applyFilters());
    }
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Filters);