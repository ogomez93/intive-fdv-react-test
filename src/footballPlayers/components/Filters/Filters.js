import React from 'react';
import { POSITIONS } from 'footballPlayers/constants';

const positions = POSITIONS.map(position =>
  <option value={ position } key={ position }>
    { position }
  </option>
);

const Filters = ({
  age,
  name,
  onAgeChange,
  onNameChange,
  onPositionChange,
  onSubmit
}) => (
  <form className="my-4 text-xs sm:flex sm:justify-between" onSubmit={ onSubmit }>
    <input
      name="playerName"
      type='text'
      className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
      onChange={ onNameChange }
      placeholder="Player Name"
      value={ name }
    />

    <select
      name="playerPosition"
      className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4 bg-white"
      onChange={ onPositionChange }
      placeholder="Position">
      <option value="">Position</option>
      { positions }
    </select>

    <input
      name="playerAge"
      className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
      type="number"
      onChange={ onAgeChange }
      placeholder="Age"
      min="18"
      max="40"
      value={ age }
    />

    <button
      className="border-2 border-black px-6 py-3 w-full sm:w-auto"
      type="submit">
      Search
    </button>
  </form>
);

export default Filters;
