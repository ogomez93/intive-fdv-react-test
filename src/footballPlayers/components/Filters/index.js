import { connect } from 'react-redux';
import { compose, withState, withHandlers } from 'recompose';

import Filters from './Filters';
import { applyFilters } from '../../actions/filters';

const mapStateToProps = ({
  footballPlayers: {
    nameFilter,
    ageFilter,
    positionFilter
  }
}) => ({
  nameFilter,
  ageFilter,
  positionFilter
});

const mapDispatchToProps = (dispatch, { name, position, age }) => ({
  onSubmit: event => {
    event.preventDefault();
    dispatch(applyFilters({ name, position, age }));
  }
});

export const enhance = compose(
  withState('name', 'setName', ''),
  withState('position', 'setPosition', ''),
  withState('age', 'setAge', ''),
  withHandlers({
    onNameChange: ({ setName }) => event =>
      setName(event.target.value),
    onPositionChange: ({ setPosition }) => event =>
      setPosition(event.target.value),
    onAgeChange: ({ setAge }) => event =>
      setAge(event.target.value)
  }),
  connect(mapStateToProps, mapDispatchToProps)
);

export default enhance(Filters);
