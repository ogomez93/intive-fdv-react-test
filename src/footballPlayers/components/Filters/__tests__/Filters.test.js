import React from 'react';
import { shallow } from 'enzyme';

import Filters from '../Filters';

const mockedOnAgeChange = jest.fn();
const mockedOnNameChange = jest.fn();
const mockedOnPositionChange = jest.fn();
const mockedOnSubmit = jest.fn();

const props = {
  age: 21,
  name: 'Test name',
  onAgeChange: mockedOnAgeChange,
  onNameChange: mockedOnNameChange,
  onPositionChange: mockedOnPositionChange,
  onSubmit: mockedOnSubmit
};

describe('Component: <Filters />', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = shallow(<Filters {...props} />);
  });

  it('should render', () => {
    expect(wrapper.isEmptyRender()).toEqual(false);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('form element', () => {
    it('should have an onSubmit handler', () => {
      expect(wrapper.prop('onSubmit')).toEqual(mockedOnSubmit);
    });
  });

  describe('name input', () => {
    let input;

    beforeAll(() => {
      input = wrapper.find('[name="playerName"]');
    });

    it('should be a text input', () => {
      expect(input.prop('type')).toEqual('text');
    });

    it('should have an onChange handler', () => {
      expect(input.prop('onChange')).toEqual(mockedOnNameChange);
    });

    it('should have a placeholder', () => {
      expect(input.prop('placeholder')).toEqual('Player Name');
    });

    it('should have a value', () => {
      expect(input.prop('value')).toEqual('Test name');
    });
  });

  describe('position select', () => {
    let select;

    beforeAll(() => {
      select = wrapper.find('[name="playerPosition"]');
    });

    it('should have an onChange handler', () => {
      expect(select.prop('onChange')).toEqual(mockedOnPositionChange);
    });

    it('should have a placeholder', () => {
      expect(select.prop('placeholder')).toEqual('Position');
    });

    it('should have an empty value as its first option', () => {
      expect(select.children().at(0).prop('value')).toEqual('');
    });
  });

  describe('age input', () => {
    let input;

    beforeAll(() => {
      input = wrapper.find('[name="playerAge"]');
    });

    it('should be a numeric input', () => {
      expect(input.prop('type')).toEqual('number');
    });

    it('should have an onChange handler', () => {
      expect(input.prop('onChange')).toEqual(mockedOnAgeChange);
    });

    it('should have a placeholder', () => {
      expect(input.prop('placeholder')).toEqual('Age');
    });

    it('should have a value', () => {
      expect(input.prop('value')).toEqual(21);
    });
  });
});
