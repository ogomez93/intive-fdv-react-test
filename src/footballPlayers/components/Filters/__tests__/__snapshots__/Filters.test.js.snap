// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Filters /> should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Filters
    age={21}
    name="Test name"
    onAgeChange={[Function]}
    onNameChange={[Function]}
    onPositionChange={[Function]}
    onSubmit={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
          name="playerName"
          onChange={[Function]}
          placeholder="Player Name"
          type="text"
          value="Test name"
/>,
        <select
          className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4 bg-white"
          name="playerPosition"
          onChange={[Function]}
          placeholder="Position"
>
          <option
                    value=""
          >
                    Position
          </option>
          <option
                    value="Attacking Midfield"
          >
                    Attacking Midfield
          </option>
          <option
                    value="Central Midfield"
          >
                    Central Midfield
          </option>
          <option
                    value="Centre-Back"
          >
                    Centre-Back
          </option>
          <option
                    value="Centre-Forward"
          >
                    Centre-Forward
          </option>
          <option
                    value="Defensive Midfield"
          >
                    Defensive Midfield
          </option>
          <option
                    value="Keeper"
          >
                    Keeper
          </option>
          <option
                    value="Left Midfield"
          >
                    Left Midfield
          </option>
          <option
                    value="Left Wing"
          >
                    Left Wing
          </option>
          <option
                    value="Left-Back"
          >
                    Left-Back
          </option>
          <option
                    value="Right-Back"
          >
                    Right-Back
          </option>
</select>,
        <input
          className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
          max="40"
          min="18"
          name="playerAge"
          onChange={[Function]}
          placeholder="Age"
          type="number"
          value={21}
/>,
        <button
          className="border-2 border-black px-6 py-3 w-full sm:w-auto"
          type="submit"
>
          Search
</button>,
      ],
      "className": "my-4 text-xs sm:flex sm:justify-between",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4",
          "name": "playerName",
          "onChange": [Function],
          "placeholder": "Player Name",
          "type": "text",
          "value": "Test name",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              value=""
>
              Position
</option>,
            Array [
              <option
                value="Attacking Midfield"
>
                Attacking Midfield
</option>,
              <option
                value="Central Midfield"
>
                Central Midfield
</option>,
              <option
                value="Centre-Back"
>
                Centre-Back
</option>,
              <option
                value="Centre-Forward"
>
                Centre-Forward
</option>,
              <option
                value="Defensive Midfield"
>
                Defensive Midfield
</option>,
              <option
                value="Keeper"
>
                Keeper
</option>,
              <option
                value="Left Midfield"
>
                Left Midfield
</option>,
              <option
                value="Left Wing"
>
                Left Wing
</option>,
              <option
                value="Left-Back"
>
                Left-Back
</option>,
              <option
                value="Right-Back"
>
                Right-Back
</option>,
            ],
          ],
          "className": "border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4 bg-white",
          "name": "playerPosition",
          "onChange": [Function],
          "placeholder": "Position",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Position",
              "value": "",
            },
            "ref": null,
            "rendered": "Position",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Attacking Midfield",
            "nodeType": "host",
            "props": Object {
              "children": "Attacking Midfield",
              "value": "Attacking Midfield",
            },
            "ref": null,
            "rendered": "Attacking Midfield",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Central Midfield",
            "nodeType": "host",
            "props": Object {
              "children": "Central Midfield",
              "value": "Central Midfield",
            },
            "ref": null,
            "rendered": "Central Midfield",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Centre-Back",
            "nodeType": "host",
            "props": Object {
              "children": "Centre-Back",
              "value": "Centre-Back",
            },
            "ref": null,
            "rendered": "Centre-Back",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Centre-Forward",
            "nodeType": "host",
            "props": Object {
              "children": "Centre-Forward",
              "value": "Centre-Forward",
            },
            "ref": null,
            "rendered": "Centre-Forward",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Defensive Midfield",
            "nodeType": "host",
            "props": Object {
              "children": "Defensive Midfield",
              "value": "Defensive Midfield",
            },
            "ref": null,
            "rendered": "Defensive Midfield",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Keeper",
            "nodeType": "host",
            "props": Object {
              "children": "Keeper",
              "value": "Keeper",
            },
            "ref": null,
            "rendered": "Keeper",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Left Midfield",
            "nodeType": "host",
            "props": Object {
              "children": "Left Midfield",
              "value": "Left Midfield",
            },
            "ref": null,
            "rendered": "Left Midfield",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Left Wing",
            "nodeType": "host",
            "props": Object {
              "children": "Left Wing",
              "value": "Left Wing",
            },
            "ref": null,
            "rendered": "Left Wing",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Left-Back",
            "nodeType": "host",
            "props": Object {
              "children": "Left-Back",
              "value": "Left-Back",
            },
            "ref": null,
            "rendered": "Left-Back",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Right-Back",
            "nodeType": "host",
            "props": Object {
              "children": "Right-Back",
              "value": "Right-Back",
            },
            "ref": null,
            "rendered": "Right-Back",
            "type": "option",
          },
        ],
        "type": "select",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4",
          "max": "40",
          "min": "18",
          "name": "playerAge",
          "onChange": [Function],
          "placeholder": "Age",
          "type": "number",
          "value": 21,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Search",
          "className": "border-2 border-black px-6 py-3 w-full sm:w-auto",
          "type": "submit",
        },
        "ref": null,
        "rendered": "Search",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
            name="playerName"
            onChange={[Function]}
            placeholder="Player Name"
            type="text"
            value="Test name"
/>,
          <select
            className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4 bg-white"
            name="playerPosition"
            onChange={[Function]}
            placeholder="Position"
>
            <option
                        value=""
            >
                        Position
            </option>
            <option
                        value="Attacking Midfield"
            >
                        Attacking Midfield
            </option>
            <option
                        value="Central Midfield"
            >
                        Central Midfield
            </option>
            <option
                        value="Centre-Back"
            >
                        Centre-Back
            </option>
            <option
                        value="Centre-Forward"
            >
                        Centre-Forward
            </option>
            <option
                        value="Defensive Midfield"
            >
                        Defensive Midfield
            </option>
            <option
                        value="Keeper"
            >
                        Keeper
            </option>
            <option
                        value="Left Midfield"
            >
                        Left Midfield
            </option>
            <option
                        value="Left Wing"
            >
                        Left Wing
            </option>
            <option
                        value="Left-Back"
            >
                        Left-Back
            </option>
            <option
                        value="Right-Back"
            >
                        Right-Back
            </option>
</select>,
          <input
            className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
            max="40"
            min="18"
            name="playerAge"
            onChange={[Function]}
            placeholder="Age"
            type="number"
            value={21}
/>,
          <button
            className="border-2 border-black px-6 py-3 w-full sm:w-auto"
            type="submit"
>
            Search
</button>,
        ],
        "className": "my-4 text-xs sm:flex sm:justify-between",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4",
            "name": "playerName",
            "onChange": [Function],
            "placeholder": "Player Name",
            "type": "text",
            "value": "Test name",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                value=""
>
                Position
</option>,
              Array [
                <option
                  value="Attacking Midfield"
>
                  Attacking Midfield
</option>,
                <option
                  value="Central Midfield"
>
                  Central Midfield
</option>,
                <option
                  value="Centre-Back"
>
                  Centre-Back
</option>,
                <option
                  value="Centre-Forward"
>
                  Centre-Forward
</option>,
                <option
                  value="Defensive Midfield"
>
                  Defensive Midfield
</option>,
                <option
                  value="Keeper"
>
                  Keeper
</option>,
                <option
                  value="Left Midfield"
>
                  Left Midfield
</option>,
                <option
                  value="Left Wing"
>
                  Left Wing
</option>,
                <option
                  value="Left-Back"
>
                  Left-Back
</option>,
                <option
                  value="Right-Back"
>
                  Right-Back
</option>,
              ],
            ],
            "className": "border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4 bg-white",
            "name": "playerPosition",
            "onChange": [Function],
            "placeholder": "Position",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Position",
                "value": "",
              },
              "ref": null,
              "rendered": "Position",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Attacking Midfield",
              "nodeType": "host",
              "props": Object {
                "children": "Attacking Midfield",
                "value": "Attacking Midfield",
              },
              "ref": null,
              "rendered": "Attacking Midfield",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Central Midfield",
              "nodeType": "host",
              "props": Object {
                "children": "Central Midfield",
                "value": "Central Midfield",
              },
              "ref": null,
              "rendered": "Central Midfield",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Centre-Back",
              "nodeType": "host",
              "props": Object {
                "children": "Centre-Back",
                "value": "Centre-Back",
              },
              "ref": null,
              "rendered": "Centre-Back",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Centre-Forward",
              "nodeType": "host",
              "props": Object {
                "children": "Centre-Forward",
                "value": "Centre-Forward",
              },
              "ref": null,
              "rendered": "Centre-Forward",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Defensive Midfield",
              "nodeType": "host",
              "props": Object {
                "children": "Defensive Midfield",
                "value": "Defensive Midfield",
              },
              "ref": null,
              "rendered": "Defensive Midfield",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Keeper",
              "nodeType": "host",
              "props": Object {
                "children": "Keeper",
                "value": "Keeper",
              },
              "ref": null,
              "rendered": "Keeper",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Left Midfield",
              "nodeType": "host",
              "props": Object {
                "children": "Left Midfield",
                "value": "Left Midfield",
              },
              "ref": null,
              "rendered": "Left Midfield",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Left Wing",
              "nodeType": "host",
              "props": Object {
                "children": "Left Wing",
                "value": "Left Wing",
              },
              "ref": null,
              "rendered": "Left Wing",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Left-Back",
              "nodeType": "host",
              "props": Object {
                "children": "Left-Back",
                "value": "Left-Back",
              },
              "ref": null,
              "rendered": "Left-Back",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Right-Back",
              "nodeType": "host",
              "props": Object {
                "children": "Right-Back",
                "value": "Right-Back",
              },
              "ref": null,
              "rendered": "Right-Back",
              "type": "option",
            },
          ],
          "type": "select",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4",
            "max": "40",
            "min": "18",
            "name": "playerAge",
            "onChange": [Function],
            "placeholder": "Age",
            "type": "number",
            "value": 21,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Search",
            "className": "border-2 border-black px-6 py-3 w-full sm:w-auto",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Search",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
